File: apps/mobile-native/composeApp/src/androidMain/kotlin/moe/ditto/halo/player/MpvCore.kt
package moe.ditto.halo.player

import android.content.Context
import android.util.Log
import android.view.Surface
import dev.jdtech.mpv.MPVLib
import java.util.concurrent.Executor
import java.util.concurrent.RejectedExecutionException

/**
 * Owned thin adapter over libmpv's Android JNI ([MPVLib], mpv-android lineage) —
 * the direct parallel to iOS's `MPVCore.swift`. The prebuilt AAR is only the
 * binary provider; MPVLib types never leave this file, so the boundary the
 * common shell sees is the platform-neutral player contract, and the binary
 * provider stays replaceable (production ships an owned reproducible build).
 *
 * mpv's own thread delivers property/events through [MPVLib.EventObserver]; this
 * class translates them to a small owned [Listener] using seconds and the
 * neutral [PlayerTracks] schema. It never surfaces an mpv event id, format int,
 * or property name upward.
 */
internal class MpvCore private constructor(
    private val mpv: MPVLib,
    val id: String,
    private val callbackExecutor: Executor?,
) {
    interface Listener {
        fun onReady(durationSeconds: Double?)
        fun onPosition(positionSeconds: Double)
        fun onPauseChanged(paused: Boolean)
        fun onTracks(tracks: PlayerTracks)
        /** Null once the cache stops stalling playback. */
        fun onBuffering(buffering: PlayerBuffering?)
        fun onBufferedPosition(positionSeconds: Double)
        fun onEnded()
        fun onError(message: String)
    }

    @Volatile private var listener: Listener? = null
    @Volatile private var destroyed = false
    // mpv fires time-pos many times per second; forward only whole-second
    // changes. The shell displays seconds, and flooding recompositions would
    // otherwise starve Compose's idle sync (and waste work).
    @Volatile private var lastPositionSecond = Long.MIN_VALUE
    // Same reasoning for the cache's reach: the transport bar draws it as a
    // fraction of a whole timeline, so sub-second updates buy nothing visible.
    @Volatile private var lastBufferedSecond = Long.MIN_VALUE
    // Whether the cache is currently stalling playback, and how full mpv
    // considers it. Held because the two arrive as separate observations and
    // the overlay needs both at once.
    @Volatile private var pausedForCache = false
    @Volatile private var cacheFillPercent: Int? = null
    // END_FILE's reason is hidden by this JNI binding. The small state machine
    // distinguishes a real failure from the END_FILE that loadfile emits for
    // the file it replaced, and serializes the caller, event and log threads.
    private val fileLifecycle = MpvFileLifecycle()

    private val eventObserver = object : MPVLib.EventObserver {
        /**
         * Value-less observations. `sid`/`aid` are observed for the fact that
         * they changed, not for what they changed to: which track is current is
         * read back off `track-list` like every other track fact, so a
         * selection mpv made on its own reports the same way an app-driven one
         * does. Without this the panel keeps highlighting the previous row
         * after a switch, because the track *count* has not moved.
         */
        override fun eventProperty(property: String) {
            dispatch { handleProperty(property) }
        }

        override fun eventProperty(property: String, value: Long) {
            dispatch { handleProperty(property, value) }
        }

        override fun eventProperty(property: String, value: Double) {
            // These observations are already values from mpv, not property
            // reads. Filter them before queueing so a high-frequency time-pos
            // stream cannot starve serialized commands behind the executor.
            when (property) {
                "time-pos" -> {
                    val second = value.toLong()
                    if (second != lastPositionSecond) {
                        lastPositionSecond = second
                        dispatch { listener?.onPosition(value) }
                    }
                }
                "demuxer-cache-time" -> {
                    val second = value.toLong()
                    if (second != lastBufferedSecond) {
                        lastBufferedSecond = second
                        dispatch { listener?.onBufferedPosition(value) }
                    }
                }
            }
        }

        override fun eventProperty(property: String, value: Boolean) {
            dispatch { handleProperty(property, value) }
        }

        override fun eventProperty(property: String, value: String) { /* unused */ }

        override fun event(eventId: Int) {
            dispatch { handleEvent(eventId) }
        }
    }

    private fun dispatch(block: () -> Unit) {
        val executor = callbackExecutor
        if (executor == null) {
            if (!destroyed) block()
            return
        }
        try {
            executor.execute {
                if (!destroyed) block()
            }
        } catch (_: RejectedExecutionException) {
            // Activity shutdown can close the executor while mpv is still
            // draining callbacks. The core is being destroyed, so dropping
            // those stale callbacks is the safe outcome.
        }
    }

    private fun handleProperty(property: String) {
        when (property) {
            "sid", "aid" -> emitTracks()
        }
    }

    private fun handleProperty(property: String, value: Long) {
        when (property) {
            "track-list/count" -> emitTracks()
            "cache-buffering-state" -> {
                cacheFillPercent = value.toInt()
                if (pausedForCache) emitBuffering()
            }
        }
    }

    private fun handleProperty(property: String, value: Boolean) {
        when (property) {
            "pause" -> listener?.onPauseChanged(value)
            "eof-reached" -> if (value && fileLifecycle.markEofReached()) {
                listener?.onEnded()
            }
            "paused-for-cache" -> {
                pausedForCache = value
                emitBuffering()
            }
        }
    }

    private fun handleEvent(eventId: Int) {
        when (eventId) {
            MPVLib.MpvEvent.MPV_EVENT_START_FILE -> fileLifecycle.onFileStarted()
            MPVLib.MpvEvent.MPV_EVENT_FILE_LOADED -> onFileLoaded()
            MPVLib.MpvEvent.MPV_EVENT_END_FILE ->
                fileLifecycle.endFileFailure()?.let { listener?.onError(it) }
            MPVLib.MpvEvent.MPV_EVENT_SHUTDOWN -> listener?.onEnded()
        }
    }

    // MPVLib.LogObserver is a plain Kotlin interface (not `fun interface`), so no
    // SAM lambda — an explicit object is required.
    private val logObserver = object : MPVLib.LogObserver {
        override fun logMessage(prefix: String, level: Int, text: String) {
            val message = text.trim()
            // Primary diagnostic channel: mpv's own log lines land in logcat under
            // a greppable tag (Configuration/vo/hwdec/subtitle-track selection).
            Log.i(LOG_TAG, "[$id][$prefix] $message")
            // Preserve exactly what mpv said. No URL, HTTP status or host is
            // added by this layer; the error card prints this string verbatim.
            if (level in MPVLib.MpvLogLevel.MPV_LOG_LEVEL_FATAL..MPVLib.MpvLogLevel.MPV_LOG_LEVEL_ERROR &&
                message.isNotEmpty()
            ) {
                fileLifecycle.recordError(message)
            }
        }
    }

    fun setListener(listener: Listener?) {
        this.listener = listener
    }

    /** Read only by Android instrumentation after [MPVLib.init] has completed. */
    fun isMutedForTest(): Boolean? = if (destroyed) null else mpv.getPropertyBoolean("mute")

    /** Attach the render surface and turn the GPU video output on (mpv-android order). */
    fun attachSurface(surface: Surface, width: Int, height: Int) {
        if (destroyed) return
        mpv.attachSurface(surface)
        mpv.setOptionString("force-window", "yes")
        mpv.setPropertyString("android-surface-size", "${width}x$height")
        mpv.setOptionString("vo", VO)
    }

    fun setSurfaceSize(width: Int, height: Int) {
        if (destroyed || width <= 0 || height <= 0) return
        mpv.setPropertyString("android-surface-size", "${width}x$height")
    }

    /**
     * Selects or drops the video track.
     *
     * Dropping it is how the decoder is destroyed on demand: mpv reinitialises
     * the video chain synchronously here, so when this returns no MediaCodec
     * instance is left holding buffers. Audio is untouched.
     */
    fun setVideoEnabled(enabled: Boolean) {
        if (destroyed) return
        mpv.setPropertyString("vid", if (enabled) "auto" else "no")
    }

    /** Turn video output off before releasing the surface (mpv-android order). */
    fun detachSurface() {
        if (destroyed) return
        mpv.setPropertyString("vo", "null")
        mpv.setOptionString("force-window", "no")
        mpv.detachSurface()
    }

    fun load(url: String) {
        if (destroyed) return
        fileLifecycle.onLoadRequested()
        mpv.command(arrayOf("loadfile", url))
    }

    fun stop() {
        if (destroyed) return
        fileLifecycle.onStopRequested()
        mpv.command(arrayOf("stop"))
    }

    fun setPaused(paused: Boolean) {
        if (destroyed) return
        mpv.setPropertyBoolean("pause", paused)
    }

    fun seekTo(positionSeconds: Double) {
        if (destroyed) return
        mpv.command(arrayOf("seek", positionSeconds.toString(), "absolute"))
    }

    fun selectAudioTrack(id: String?) {
        if (destroyed) return
        mpv.setPropertyString("aid", id ?: "no")
    }

    fun selectSubtitleTrack(id: String?) {
        if (destroyed) return
        mpv.setPropertyString("sid", id ?: "no")
    }

    fun setPlaybackRate(rate: Double) {
        if (destroyed) return
        mpv.setPropertyDouble("speed", rate)
    }

    /**
     * `panscan` is a 0..1 blend between fitting the picture and cropping it to
     * the screen, and the two ends are the only settings the design offers.
     */
    fun setVideoFillsScreen(fills: Boolean) {
        if (destroyed) return
        mpv.setPropertyDouble("panscan", if (fills) 1.0 else 0.0)
    }

    fun setAudioDelay(seconds: Double) {
        if (destroyed) return
        mpv.setPropertyDouble("audio-delay", seconds)
    }

    // Live subtitle controls — the exact capability libVLC lacked on mobile:
    // these apply to the running core with no recreation.
    fun setSubtitleDelay(seconds: Double) {
        if (destroyed) return
        mpv.setPropertyDouble("sub-delay", seconds)
    }

    fun setSubtitleScale(scale: Double) {
        if (destroyed) return
        mpv.setPropertyDouble("sub-scale", scale)
    }

    fun setSubtitleFont(font: String?) {
        if (destroyed) return
        mpv.setPropertyString("sub-font", font ?: DefaultSubtitleFont)
    }

    /**
     * `no` leaves a script's own styling alone; `force` replaces its fonts,
     * sizes and positions with the player's. mpv has no third state, and the
     * two names read backwards from the switch, which is why the mapping is
     * here rather than at the call site.
     */
    fun setSubtitleTrackStyling(keepScript: Boolean) {
        if (destroyed) return
        mpv.setPropertyString("sub-ass-override", if (keepScript) "no" else "force")
    }

    fun setSubtitleOutline(widthPixels: Double) {
        if (destroyed) return
        mpv.setPropertyDouble("sub-border-size", widthPixels)
    }

    fun setSubtitleShadow(offsetPixels: Double) {
        if (destroyed) return
        mpv.setPropertyDouble("sub-shadow-offset", offsetPixels)
    }

    /**
     * mpv counts `sub-pos` downwards from the top, so a lift is a subtraction.
     * 100 is the resting place at the bottom edge.
     */
    fun setSubtitleLift(percent: Int) {
        if (destroyed) return
        mpv.setPropertyInt("sub-pos", 100 - percent)
    }

    /**
     * `cached` rather than `select`: it re-selects a URL already added instead
     * of adding it a second time. Restoring a remembered subtitle and then
     * tapping the same row otherwise leaves two identical tracks in the list,
     * which is visible on device and cannot be undone from the UI.
     */
    fun addSubtitle(url: String) {
        if (destroyed) return
        mpv.command(arrayOf("sub-add", url, "cached"))
    }

    /**
     * Deterministic, idempotent teardown (mirrors the iOS core's contract):
     * detach observers first so no event races the native destroy, then
     * terminate. Safe to call twice.
     */
    fun destroy() {
        if (destroyed) return
        destroyed = true
        listener = null
        runCatching { mpv.removeObserver(eventObserver) }
        runCatching { mpv.removeLogObserver(logObserver) }
        runCatching { mpv.destroy() }
    }

    private fun onFileLoaded() {
        val duration = mpv.getPropertyDouble("duration")?.takeIf { it.isFinite() && it >= 0.0 }
        listener?.onReady(duration)
        emitTracks()
    }

    /**
     * Reports the stall, reading the rate and depth only while one is happening.
     *
     * `paused-for-cache` is the gate rather than `cache-buffering-state`,
     * because the latter sits below 100 through perfectly healthy streaming and
     * would flash the overlay over a picture that never stopped moving. The
     * two figures are read here instead of observed: they only matter while the
     * overlay is up, and mpv revises the fill percentage as it refills, so this
     * runs again for each revision.
     */
    private fun emitBuffering() {
        if (destroyed) return
        if (!pausedForCache) {
            listener?.onBuffering(null)
            return
        }
        listener?.onBuffering(
            PlayerBuffering(
                percent = cacheFillPercent,
                // mpv types cache-speed as int64; the JNI hands back an Int,
                // which cannot hold a rate a phone will ever see anyway.
                bytesPerSecond = mpv.getPropertyInt("cache-speed")?.toLong(),
                cachedSeconds = mpv.getPropertyDouble("demuxer-cache-duration"),
            ),
        )
    }

    private fun emitTracks() {
        if (destroyed) return
        val count = mpv.getPropertyInt("track-list/count") ?: return
        val audio = mutableListOf<PlayerTrack>()
        val subs = mutableListOf<PlayerTrack>()
        var selectedAudio: String? = null
        var selectedSub: String? = null
        for (i in 0 until count) {
            val type = mpv.getPropertyString("track-list/$i/type") ?: continue
            val trackId = mpv.getPropertyInt("track-list/$i/id")?.toString() ?: continue
            val lang = mpv.getPropertyString("track-list/$i/lang")
            val title = mpv.getPropertyString("track-list/$i/title")
            val selected = mpv.getPropertyBoolean("track-list/$i/selected") ?: false
            val codec = mpv.getPropertyString("track-list/$i/codec")
            val channels = mpv.getPropertyInt("track-list/$i/demux-channel-count")
            val sampleRateHz = mpv.getPropertyInt("track-list/$i/demux-samplerate")
            val label = title ?: lang ?: "$type $trackId"
            when (type) {
                "audio" -> {
                    audio += PlayerTrack(trackId, label, lang, codec, channels, sampleRateHz)
                    if (selected) selectedAudio = trackId
                }
                "sub" -> {
                    subs += PlayerTrack(trackId, label, lang, codec = codec)
                    if (selected) selectedSub = trackId
                }
            }
        }
        listener?.onTracks(
            PlayerTracks(
                audio = audio,
                subtitles = subs,
                selectedAudioId = selectedAudio,
                selectedSubtitleId = selectedSub,
            ),
        )
    }

    companion object {
        const val LOG_TAG = "HALO_MPV"
        private const val VO = "gpu"

        /** What "no particular font" means to the renderer. */
        private const val DefaultSubtitleFont = "sans-serif"

        /** Create + configure + initialize a fresh core with a stable [id]. */
        fun create(context: Context, id: String, callbackExecutor: Executor? = null): MpvCore {
            val mpv = MPVLib.create(context) ?: error("MPVLib.create returned null")
            // Pre-init options (mpv-android's proven Android render/decode stack).
            mpv.setOptionString("config", "no")
            mpv.setOptionString("vo", "null") // no output until a surface attaches
            mpv.setOptionString("gpu-context", "android")
            mpv.setOptionString("opengl-es", "yes")
            mpv.setOptionString("hwdec", "mediacodec-copy") // auto-falls back to sw
            mpv.setOptionString("ao", "audiotrack")
            mpv.setOptionString("keep-open", "yes") // so eof-reached fires
            // mpv's default is to hand a URL it could not open to youtube-dl.
            // There is no such binary on Android and never will be, so every
            // failed open would otherwise spawn a subprocess, fail, and bury the
            // real diagnostic under its own error. The error card prints the
            // engine's message verbatim, so that noise is not free.
            mpv.setOptionString("ytdl", "no")
            // Subtitle auto-select + rendering; embedded fonts make ASS render
            // without depending on Android system fonts.
            mpv.setOptionString("slang", "eng,en")
            mpv.setOptionString("subs-fallback", "yes")
            mpv.setOptionString("embeddedfonts", "yes")
            // Fonts the app ships, so a chosen family resolves to that family
            // rather than to whichever system face fontconfig settles on.
            // Pre-init: libass builds its font provider during initialisation.
            SubtitleFontLibrary.prepare(context)?.let { mpv.setOptionString("sub-fonts-dir", it) }

            val core = MpvCore(mpv, id, callbackExecutor)
            mpv.addLogObserver(core.logObserver)
            mpv.addObserver(core.eventObserver)
            mpv.init()

            // Observe the neutral property set the boundary needs.
            mpv.observeProperty("time-pos", MPVLib.MpvFormat.MPV_FORMAT_DOUBLE)
            mpv.observeProperty("pause", MPVLib.MpvFormat.MPV_FORMAT_FLAG)
            mpv.observeProperty("eof-reached", MPVLib.MpvFormat.MPV_FORMAT_FLAG)
            mpv.observeProperty("track-list/count", MPVLib.MpvFormat.MPV_FORMAT_INT64)
            // Selection, which the count cannot report. Same pair the iOS core
            // observes, so both platforms answer "which track is playing" from
            // the engine rather than from what the UI last asked for.
            mpv.observeProperty("sid", MPVLib.MpvFormat.MPV_FORMAT_NONE)
            mpv.observeProperty("aid", MPVLib.MpvFormat.MPV_FORMAT_NONE)
            mpv.observeProperty("paused-for-cache", MPVLib.MpvFormat.MPV_FORMAT_FLAG)
            mpv.observeProperty("cache-buffering-state", MPVLib.MpvFormat.MPV_FORMAT_INT64)
            mpv.observeProperty("demuxer-cache-time", MPVLib.MpvFormat.MPV_FORMAT_DOUBLE)
            return core
        }
    }
}

/**
 * Attributes mpv file-ending events when the Android binding exposes only the
 * event id, not `mpv_end_file_reason`.
 *
 * `loadfile` ends the old file before starting its replacement. That expected
 * END_FILE must not fail the new item. The counter, rather than a Boolean,
 * also handles two rapid replacements before mpv's event thread catches up.
 */
internal class MpvFileLifecycle {
    private val lock = Any()
    private var fileActive = false
    private var expectedReplacementEndFiles = 0
    private var lastErrorMessage: String? = null
    private var eofReached = false

    fun onLoadRequested() = synchronized(lock) {
        if (fileActive) expectedReplacementEndFiles += 1
        fileActive = true
        lastErrorMessage = null
        eofReached = false
    }

    fun onStopRequested() = synchronized(lock) {
        if (!fileActive) return@synchronized
        expectedReplacementEndFiles += 1
        fileActive = false
    }

    /**
     * START_FILE separates diagnostics from the replaced connection and the
     * new one. Harmless TLS close noise from the old stream must not be shown
     * if the new stream later has a genuine failure.
     */
    fun onFileStarted() = synchronized(lock) {
        lastErrorMessage = null
        eofReached = false
    }

    /** Keeps the first useful diagnostic because later hooks are often generic. */
    fun recordError(message: String) = synchronized(lock) {
        if (lastErrorMessage == null) lastErrorMessage = message
    }

    /** True only for the first EOF observation for the current file. */
    fun markEofReached(): Boolean = synchronized(lock) {
        if (eofReached) return@synchronized false
        eofReached = true
        true
    }

    /** A real current-file failure, or null for replacement and natural endings. */
    fun endFileFailure(): String? = synchronized(lock) {
        if (expectedReplacementEndFiles > 0) {
            expectedReplacementEndFiles -= 1
            return@synchronized null
        }
        fileActive = false
        if (eofReached) return@synchronized null
        lastErrorMessage ?: "Playback ended unexpectedly."
    }
}

File: apps/mobile-native/iosApp/Halo/MPVCore.swift
import Foundation
import UIKit
import AVFAudio
import Libmpv

/// Normalized playback events. `tracksJson` is the neutral schema the Kotlin
/// side parses (`PlayerTracksJson`) — mpv's own `track-list` shape stays here.
enum MPVCoreEvent {
    case ready(durationSeconds: Double)
    case position(seconds: Double)
    case pauseChanged(paused: Bool)
    case tracks(json: String)
    case ended
    case error(message: String)
}

/// One owned libmpv instance rendering via gpu-next/MoltenVK into a
/// `MetalLayer` handed to mpv as `wid`.
///
/// Ownership and threading rules:
/// - Every mpv call happens on the private serial `queue`; the wakeup callback
///   only schedules a drain onto that queue and never touches mpv itself.
/// - `shutdown()` is deterministic and idempotent: it clears the wakeup
///   callback, then runs `mpv_terminate_destroy` on the queue and blocks until
///   it returns. libmpv guarantees no callback runs after terminate_destroy
///   returns, so the owner may release the core immediately afterwards.
/// - After shutdown every command and event is rejected; an
///   unretained-callback / teardown-less lifecycle is deliberately not
///   supported.
final class MPVCore {
    let id = UUID().uuidString
    /// Owned by this core: a fresh layer per core keeps mpv's `wid` pointing
    /// at a surface whose lifetime exactly matches the mpv handle.
    let videoLayer = MetalLayer()

    private var mpv: OpaquePointer?
    private let queue = DispatchQueue(label: "halo-mpv-core", qos: .userInitiated)
    // Everything below is only touched on `queue` after init — including
    // `onEvent`: a fresh core emits initial property observations the moment
    // the wakeup callback registers, so a main-thread assignment would race
    // `emit()` invoking the closure on the queue (caught by TSan as a SEGV in
    // the released closure's context during recreate cycles). Handlers are
    // installed via `setEventHandler`, which hops onto the queue; events that
    // fire before the handler lands are dropped, matching the presenter's
    // Idle guard for initial observation spam.
    private var onEvent: ((MPVCoreEvent) -> Void)?
    private var isShutdown = false
    private var eofNotified = false

    private enum Observed: UInt64 {
        case timePos = 1
        case pause = 2
        case eofReached = 3
        case trackList = 4
        case audioId = 5
        case subtitleId = 6
    }

    init(muted: Bool) {
        guard let handle = mpv_create() else {
            fatalError("mpv_create failed")
        }
        mpv = handle

        check(mpv_request_log_messages(handle, "warn"))
        // Diagnostics: full mpv log per core. On the simulator /tmp is
        // the host's /tmp, so the file is directly readable on the Mac.
        check(mpv_set_option_string(handle, "log-file", "/tmp/halo-mpv-\(id).log"))

        // Output path: layer pointer as wid, gpu-next over
        // Vulkan/MoltenVK, VideoToolbox hardware decode.
        var layerRef = Int64(Int(bitPattern: Unmanaged.passUnretained(videoLayer).toOpaque()))
        check(mpv_set_option(handle, "wid", MPV_FORMAT_INT64, &layerRef))
        check(mpv_set_option_string(handle, "vo", "gpu-next"))
        check(mpv_set_option_string(handle, "gpu-api", "vulkan"))
        check(mpv_set_option_string(handle, "gpu-context", "moltenvk"))
        check(mpv_set_option_string(handle, "hwdec", "videotoolbox"))
        check(mpv_set_option_string(handle, "keep-open", "yes"))
        // Match the desktop client's subtitle defaults so behavior carries over.
        check(mpv_set_option_string(handle, "sub-auto", "no"))
        check(mpv_set_option_string(handle, "slang", "eng,en"))
        check(mpv_set_option_string(handle, "subs-fallback", "yes"))
        if muted {
            check(mpv_set_option_string(handle, "mute", "yes"))
        }
        check(mpv_initialize(handle))

        mpv_observe_property(handle, Observed.timePos.rawValue, "time-pos", MPV_FORMAT_DOUBLE)
        mpv_observe_property(handle, Observed.pause.rawValue, "pause", MPV_FORMAT_FLAG)
        mpv_observe_property(handle, Observed.eofReached.rawValue, "eof-reached", MPV_FORMAT_FLAG)
        mpv_observe_property(handle, Observed.trackList.rawValue, "track-list", MPV_FORMAT_NONE)
        mpv_observe_property(handle, Observed.audioId.rawValue, "aid", MPV_FORMAT_NONE)
        mpv_observe_property(handle, Observed.subtitleId.rawValue, "sid", MPV_FORMAT_NONE)

        // The wakeup callback runs on an mpv-internal thread. It must only
        // schedule a drain; the unretained pointer is safe because shutdown()
        // blocks on mpv_terminate_destroy before the owner can release us.
        mpv_set_wakeup_callback(handle, { ctx in
            guard let ctx else { return }
            let core = Unmanaged<MPVCore>.fromOpaque(ctx).takeUnretainedValue()
            core.scheduleDrain()
        }, Unmanaged.passUnretained(self).toOpaque())
    }

    /// Installs the event handler on the core queue, serialized with `emit`.
    /// Never assign the handler from another thread directly.
    func setEventHandler(_ handler: @escaping (MPVCoreEvent) -> Void) {
        queue.async { [weak self] in
            guard let self, !self.isShutdown else { return }
            self.onEvent = handler
        }
    }

    // MARK: playback surface commands (all async onto the core queue)

    /// Loads start paused: deterministic for automation (no wall-clock playback
    /// racing UI assertions) and no surprise streaming of the 700 MB sample.
    func load(url: String) {
        perform { handle in
            self.eofNotified = false
            self.setFlag(handle, "pause", true)
            self.command(handle, "loadfile", [url, "replace"])
        }
    }

    func setPaused(_ paused: Bool) {
        perform { handle in self.setFlag(handle, "pause", paused) }
    }

    func seekTo(seconds: Double) {
        perform { handle in self.command(handle, "seek", [String(seconds), "absolute"]) }
    }

    func selectTrack(kind: String, id: String?) {
        perform { handle in
            mpv_set_property_string(handle, kind, id ?? "no")
        }
    }

    func setSubtitleDelay(seconds: Double) {
        perform { handle in mpv_set_property_string(handle, "sub-delay", String(seconds)) }
    }

    func setSubtitleScale(_ scale: Double) {
        perform { handle in mpv_set_property_string(handle, "sub-scale", String(scale)) }
    }

    func setSubtitleFont(_ font: String?) {
        // "sans-serif" is mpv's documented sub-font default.
        perform { handle in mpv_set_property_string(handle, "sub-font", font ?? "sans-serif") }
    }

    func addSubtitle(url: String) {
        perform { handle in self.command(handle, "sub-add", [url, "select"]) }
    }

    func setVideoDecodingSuspended(_ suspended: Bool) {
        // Background probe: drop the video track so GPU work
        // stops (mandatory in the background) while audio keeps playing.
        perform { handle in mpv_set_property_string(handle, "vid", suspended ? "no" : "auto") }
    }

    // MARK: shutdown

    /// Blocks until the mpv core is fully destroyed. Safe to call repeatedly.
    func shutdown() {
        queue.sync {
            guard let handle = mpv, !isShutdown else { return }
            isShutdown = true
            onEvent = nil
            // Stop new wakeups, then terminate. mpv_terminate_destroy joins
            // mpv's threads; when it returns no callback can run again.
            mpv_set_wakeup_callback(handle, nil, nil)
            mpv_unobserve_property(handle, Observed.timePos.rawValue)
            mpv_unobserve_property(handle, Observed.pause.rawValue)
            mpv_unobserve_property(handle, Observed.eofReached.rawValue)
            mpv_unobserve_property(handle, Observed.trackList.rawValue)
            mpv_unobserve_property(handle, Observed.audioId.rawValue)
            mpv_unobserve_property(handle, Observed.subtitleId.rawValue)
            mpv_terminate_destroy(handle)
            mpv = nil
        }
    }

    // MARK: internals (queue only)

    private func perform(_ body: @escaping (OpaquePointer) -> Void) {
        queue.async { [weak self] in
            guard let self, let handle = self.mpv, !self.isShutdown else { return }
            body(handle)
        }
    }

    private func scheduleDrain() {
        queue.async { [weak self] in self?.drainEvents() }
    }

    private func drainEvents() {
        while let handle = mpv, !isShutdown {
            guard let event = mpv_wait_event(handle, 0),
                  event.pointee.event_id != MPV_EVENT_NONE else { return }
            handleEvent(handle, event.pointee)
        }
    }

    private func handleEvent(_ handle: OpaquePointer, _ event: mpv_event) {
        switch event.event_id {
        case MPV_EVENT_FILE_LOADED:
            let duration = getDouble(handle, "duration")
            emit(.ready(durationSeconds: duration > 0 ? duration : -1))
            // Ready forces the presenter to Playing; reassert the real pause
            // state (loads start paused) right after it.
            emit(.pauseChanged(paused: getFlag(handle, "pause")))
            emitTracks(handle)
        case MPV_EVENT_END_FILE:
            guard let data = event.data else { return }
            let endFile = data.assumingMemoryBound(to: mpv_event_end_file.self).pointee
            if endFile.reason == MPV_END_FILE_REASON_ERROR {
                emit(.error(message: String(cString: mpv_error_string(endFile.error))))
            }
        case MPV_EVENT_PROPERTY_CHANGE:
            handlePropertyChange(handle, event)
        case MPV_EVENT_LOG_MESSAGE:
            if let data = event.data {
                let msg = data.assumingMemoryBound(to: mpv_event_log_message.self).pointee
                if let level = msg.level, let text = msg.text {
                    print("[mpv \(String(cString: level))] \(String(cString: text))", terminator: "")
                }
            }
        default:
            break
        }
    }

    private func handlePropertyChange(_ handle: OpaquePointer, _ event: mpv_event) {
        guard let data = event.data else { return }
        let property = data.assumingMemoryBound(to: mpv_event_property.self).pointee
        switch Observed(rawValue: event.reply_userdata) {
        case .timePos:
            guard property.format == MPV_FORMAT_DOUBLE, let value = property.data else { return }
            emit(.position(seconds: value.assumingMemoryBound(to: Double.self).pointee))
        case .pause:
            guard property.format == MPV_FORMAT_FLAG, let value = property.data else { return }
            emit(.pauseChanged(paused: value.assumingMemoryBound(to: Int32.self).pointee != 0))
        case .eofReached:
            guard property.format == MPV_FORMAT_FLAG, let value = property.data else { return }
            let reached = value.assumingMemoryBound(to: Int32.self).pointee != 0
            if reached && !eofNotified {
                eofNotified = true
                emit(.ended)
            } else if !reached {
                eofNotified = false
            }
        case .trackList, .audioId, .subtitleId:
            emitTracks(handle)
        case nil:
            break
        }
    }

    /// Normalizes mpv's `track-list` into the boundary's neutral JSON schema.
    private func emitTracks(_ handle: OpaquePointer) {
        guard let raw = mpv_get_property_string(handle, "track-list") else { return }
        defer { mpv_free(raw) }
        let json = String(cString: raw)
        guard let data = json.data(using: .utf8),
              let entries = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return
        }

        var audio: [[String: Any]] = []
        var subtitles: [[String: Any]] = []
        var selectedAudioId: String?
        var selectedSubtitleId: String?
        for entry in entries {
            guard let type = entry["type"] as? String, let id = entry["id"] as? Int else { continue }
            let language = entry["lang"] as? String
            let title = entry["title"] as? String
            var track: [String: Any] = [
                "id": String(id),
                "label": title ?? language ?? "Track \(id)",
            ]
            if let language { track["language"] = language }
            let selected = entry["selected"] as? Bool ?? false
            if type == "audio" {
                audio.append(track)
                if selected { selectedAudioId = String(id) }
            } else if type == "sub" {
                subtitles.append(track)
                if selected { selectedSubtitleId = String(id) }
            }
        }

        var document: [String: Any] = ["audio": audio, "subtitles": subtitles]
        if let selectedAudioId { document["selectedAudioId"] = selectedAudioId }
        if let selectedSubtitleId { document["selectedSubtitleId"] = selectedSubtitleId }
        guard let normalized = try? JSONSerialization.data(withJSONObject: document),
              let normalizedJson = String(data: normalized, encoding: .utf8) else {
            return
        }
        emit(.tracks(json: normalizedJson))
    }

    private func emit(_ event: MPVCoreEvent) {
        guard !isShutdown else { return }
        onEvent?(event)
    }

    // MARK: mpv helpers (queue only)

    private func command(_ handle: OpaquePointer, _ name: String, _ args: [String]) {
        var cargs: [UnsafePointer<CChar>?] = ([name] + args).map { UnsafePointer(strdup($0)) }
        cargs.append(nil)
        defer { cargs.compactMap { $0 }.forEach { free(UnsafeMutablePointer(mutating: $0)) } }
        check(mpv_command(handle, &cargs))
    }

    private func setFlag(_ handle: OpaquePointer, _ name: String, _ value: Bool) {
        var flag: Int32 = value ? 1 : 0
        mpv_set_property(handle, name, MPV_FORMAT_FLAG, &flag)
    }

    private func getFlag(_ handle: OpaquePointer, _ name: String) -> Bool {
        var flag: Int32 = 0
        mpv_get_property(handle, name, MPV_FORMAT_FLAG, &flag)
        return flag != 0
    }

    private func getDouble(_ handle: OpaquePointer, _ name: String) -> Double {
        var value = Double(0)
        mpv_get_property(handle, name, MPV_FORMAT_DOUBLE, &value)
        return value
    }

    private func check(_ status: CInt) {
        if status < 0 {
            print("mpv error: \(String(cString: mpv_error_string(status)))")
        }
    }
}
